service/workflow/models.py

```diff
@@ -1,280 +1,294 @@
-        self.validate_net()
-
-        while True:
-
-            fireable_transitions = self.get_fireable_transitions()
-            if not fireable_transitions:
-                print('\n###########################################################################')
-                print('############################## FIRING STALLED #############################')
-                print('###########################################################################\n')
-                break
-
-            # All enabled transitions must fire during each
-            # round to prevent corruption of net state.
-            with transaction.atomic():
-
-                sinks = []
-                stale_tokens = []
-
-                # Fire the transitions
-                for transition in fireable_transitions:
-                    print(f'\n****************************************************************')
-                    print(f'Evaluating at transition {transition}...')
-                    print(f'################################################################\n')
-                    for case in transition.fireable_cases():
-                        print(f'Evaluating case {case}')
-                        tokens = []
-                        for token in transition.all_tokens():
-                            if token.case is not None:
-                                if token.case.id == case.id:
-                                    print(token)
-                                    print(token.__dict__)
-                                    tokens.append(token)
-                        print(tokens)
 
-                        try:
-                            sink = transition.fire([token.__dict__ for token in tokens], case)
-                        except Exception as e:
-                            print("transition did not fire")
-                            raise e
-
-                        print(f'\nNew Tokens: {sink}')
-                        # Mark tokens as stale as transitions that depend on them fire.
-                        # Defer deletion until after this round of transitions have fired.
-                        stale_tokens = list(chain(stale_tokens, tokens))
-                        # Defer creation until after this round of transitions have fired.
-                        #import ipdb; ipdb.set_trace()
-                        sinks = list(chain(sinks, sink))
-
-                # All transitions have fired, delete stale tokens
-                print(f'\nDeleting stale tokens')
-                for token in stale_tokens:
-                    print(f'Token {token}')
-                    token.delete()
-
-                # All transitions have fired, now create tokens
-                print('\nSaving newly created tokens...')
-                for sink in sinks:
-                    print(f'Token {sink}')
-                    Token(**sink).save()
```

Comment:
Here tokens are deleted and then recreated separately from when the transitions fire, so that for each iteration any transitions that fire all see the same state. This is missing from the new scheduler and will need to be re added.